File: VodArchiver/VodArchiver/VideoJobs/FFMpegSplitJob.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Xml;
using VodArchiver.VideoInfo;

namespace VodArchiver.VideoJobs {
	class FFMpegSplitJob : IVideoJob {
		private string SplitTimes;

		public FFMpegSplitJob( string path, string splitTimes, StatusUpdate.IStatusUpdate statusUpdater = null ) {
			JobStatus = VideoJobStatus.NotStarted;
			StatusUpdater = statusUpdater == null ? new StatusUpdate.NullStatusUpdate() : statusUpdater;
			VideoInfo = new GenericVideoInfo() { Service = StreamService.FFMpegJob, VideoId = path, Username = "_split", VideoTitle = splitTimes };
			Status = "...";
			SplitTimes = splitTimes;
		}

		public FFMpegSplitJob( XmlNode node ) : base( node ) {
			SplitTimes = node.Attributes["splitTimes"].Value;
			if ( VideoInfo is GenericVideoInfo ) {
				VideoInfo.VideoTitle = SplitTimes;
			}
		}

		public override bool IsWaitingForUserInput => false;

		public override XmlNode Serialize( XmlDocument document, XmlNode node ) {
			node.AppendAttribute( document, "_type", "FFMpegSplitJob" );
			XmlNode n = base.Serialize( document, node );
			n.AppendAttribute( document, "splitTimes", SplitTimes );
			return n;
		}

		private string GenerateOutputName(string inputName) {
			string dir = System.IO.Path.GetDirectoryName(inputName);
			return System.IO.Path.Combine(dir, GenerateOutputFilename());
		}

		public override string GenerateOutputFilename() {
			string inputName = Path.GetFileName(VideoInfo.VideoId);
			string fn = System.IO.Path.GetFileNameWithoutExtension(inputName);
			string ext = System.IO.Path.GetExtension(inputName);
			var split = fn.Split(new char[] { '_' }, StringSplitOptions.None);
			List<string> output = new List<string>();

			bool added = false;
			foreach (string s in split) {
				ulong val;
				if (!added && s.StartsWith("v") && ulong.TryParse(s.Substring(1), out val)) {
					output.Add(s + "-p%d");
					added = true;
				} else {
					output.Add(s);
				}
			}

			if (!added) {
				output.Add("%d");
			}

			return string.Join("_", output) + ext;
		}

		public override async Task<ResultType> Run( CancellationToken cancellationToken ) {
			JobStatus = VideoJobStatus.Running;
			Status = "Checking files...";

			if ( cancellationToken.IsCancellationRequested ) { return ResultType.Cancelled; }

			string originalpath = Path.GetFullPath( VideoInfo.VideoId );
			string newdirpath = Path.Combine(Path.GetDirectoryName(originalpath), DateTime.UtcNow.ToString("yyyyMMddHHmmssfff"));
			if ( File.Exists( newdirpath ) || Directory.Exists( newdirpath ) ) {
				Status = "File or directory at " + newdirpath + " already exists, cancelling.";
				return ResultType.Failure;
			}
			DirectoryInfo di = Directory.CreateDirectory( newdirpath );
			string inname = Path.Combine( di.FullName, Path.GetFileName( originalpath ) );
			if ( File.Exists( inname ) ) {
				Status = "File at " + inname + " already exists, cancelling.";
				return ResultType.Failure;
			}
			File.Move( originalpath, inname );

			string outname = GenerateOutputName( inname );

			List<string> args = new List<string>();
			args.Add( "-i" );
			args.Add( inname );
			args.Add( "-codec" );
			args.Add( "copy" );
			args.Add( "-map" );
			args.Add( "0" );
			args.Add( "-f" );
			args.Add( "segment" );
			args.Add( "-segment_times" );
			args.Add( SplitTimes );
			args.Add( outname );
			args.Add( "-start_number" );
			args.Add( "1" );

			string exampleOutname = outname.Replace( "%d", "0" );
			bool existedBefore = File.Exists( exampleOutname );

			try {
				await Util.ExpensiveDiskIOSemaphore.WaitAsync();
			} catch ( OperationCanceledException ) {
				Status = "Cancelled";
				return ResultType.Cancelled;
			}
			try {
				Status = "Splitting...";
				await StallWrite( exampleOutname, new FileInfo( inname ).Length, cancellationToken );
				await VodArchiver.ExternalProgramExecution.RunProgram( "ffmpeg_split", args.ToArray() );
			} finally {
				Util.ExpensiveDiskIOSemaphore.Release();
			}


			if ( !existedBefore && File.Exists( exampleOutname ) ) {
				// output generated with indices starting at 0 instead of 1, rename
				List<(string input, string output)> l = new List<(string input, string output)>();
				int digit = 0;
				while ( true ) {
					string input = outname.Replace( "%d", digit.ToString() );
					string output = outname.Replace( "%d", ( digit + 1 ).ToString() );
					if ( File.Exists( input ) ) {
						l.Add( (input, output) );
					} else {
						break;
					}
					++digit;
				}

				l.Reverse();

				foreach ( var (input, output) in l ) {
					if ( !File.Exists( output ) ) {
						File.Move( input, output );
					}
				}
			}

			Status = "Done!";
			JobStatus = VideoJobStatus.Finished;
			return ResultType.Success;
		}
	}
}

File: native/vodarchiver/videojobs/ffmpeg-split-job.cpp
#include "ffmpeg-split-job.h"

#include <format>
#include <memory>
#include <string>
#include <string_view>
#include <utility>

#include "i-video-job.h"
#include "util/file.h"
#include "util/number.h"
#include "util/text.h"

#include "vodarchiver/exec.h"
#include "vodarchiver/videoinfo/generic-video-info.h"

namespace VodArchiver {
FFMpegSplitJob::FFMpegSplitJob() {}

FFMpegSplitJob::FFMpegSplitJob(std::string path, std::string splitTimes) {
    JobStatus = VideoJobStatus::NotStarted;
    auto vi = std::make_unique<GenericVideoInfo>();
    vi->Service = StreamService::FFMpegJob;
    vi->VideoId = std::move(path);
    vi->Username = "_split";
    vi->VideoTitle = splitTimes;
    VideoInfo = std::move(vi);
    TextStatus = "...";
    SplitTimes = std::move(splitTimes);
}

bool FFMpegSplitJob::IsWaitingForUserInput() const {
    return false;
}

static std::string GenerateOutputFilenameInternal(std::string_view fullPath) {
    std::string_view inputName =
        HyoutaUtils::TextUtils::Replace(HyoutaUtils::IO::SplitPath(fullPath).Filename, "%", "%%");
    std::string_view fn = HyoutaUtils::IO::GetFileNameWithoutExtension(inputName);
    std::string_view ext = HyoutaUtils::IO::GetExtension(inputName);
    std::vector<std::string_view> split = HyoutaUtils::TextUtils::Split(fn, "_");
    std::vector<std::string> output;

    bool added = false;
    for (const std::string_view& s : split) {
        if (!added && s.starts_with("v")
            && HyoutaUtils::NumberUtils::ParseUInt64(s.substr(1)).has_value()) {
            output.push_back(std::string(s) + "-p%d");
            added = true;
        } else {
            output.push_back(std::string(s));
        }
    }

    if (!added) {
        output.push_back("%d");
    }

    std::string rejoined = HyoutaUtils::TextUtils::Join(output, "_");
    rejoined.append(ext);
    return rejoined;
}

static std::string GenerateOutputName(std::string_view inputName, std::string_view fullPath) {
    std::string dir = std::string(HyoutaUtils::IO::SplitPath(inputName).Directory);
    HyoutaUtils::IO::AppendPathElement(dir, GenerateOutputFilenameInternal(fullPath));
    return dir;
}

static std::string EvaluateOutputName(std::string_view input, size_t index) {
    std::string result;
    size_t i = 0;
    while (i < input.size()) {
        const char ch = input[i];
        if (ch == '%') {
            ++i;
            if (i < input.size()) {
                const char format = input[i];
                if (format == '%') {
                    // %% -> %
                    result.push_back('%');
                    ++i;
                    continue;
                }
                if (format == 'd') {
                    // %d -> index
                    result.append(std::format("{}", index));
                    ++i;
                    continue;
                }

                // unclear what to do here, should probably report error?
                result.push_back(ch);
                result.push_back(format);
                ++i;
                continue;
            } else {
                // unclear what to do here, should probably report error?
                result.push_back(ch);
                break;
            }
        } else {
            result.push_back(ch);
            ++i;
        }
    }
    return result;
}

static ResultType
    RunSplitJob(FFMpegSplitJob& job, JobConfig& jobConfig, TaskCancellation& cancellationToken) {
    std::unique_ptr<IVideoInfo> videoInfo;
    std::string splitTimes;
    {
        std::lock_guard lock(*jobConfig.JobsLock);
        job.JobStatus = VideoJobStatus::Running;
        job.TextStatus = "Checking files...";
        videoInfo = job.VideoInfo->Clone();
        splitTimes = job.SplitTimes;
    }

    if (cancellationToken.IsCancellationRequested()) {
        return ResultType::Cancelled;
    }

    std::array<char, 256> buffer;
    std::string originalpath = HyoutaUtils::IO::GetAbsolutePath(videoInfo->GetVideoId(buffer));
    std::string newdirpath(HyoutaUtils::IO::GetDirectoryName(originalpath));
    HyoutaUtils::IO::AppendPathElement(newdirpath,
                                       std::format("{}", DateTime::UtcNow().GetTicks()));
    if (HyoutaUtils::IO::Exists(std::string_view(newdirpath))
        == HyoutaUtils::IO::ExistsResult::DoesExist) {
        std::lock_guard lock(*jobConfig.JobsLock);
        job.TextStatus =
            std::format("File or directory at {} already exists, cancelling.", newdirpath);
        return ResultType::Failure;
    }
    if (!HyoutaUtils::IO::CreateDirectory(std::string_view(newdirpath))) {
        std::lock_guard lock(*jobConfig.JobsLock);
        job.TextStatus = std::format("Failed to create directory at {}, cancelling.", newdirpath);
        return ResultType::Failure;
    }
    std::string inname = newdirpath;
    HyoutaUtils::IO::AppendPathElement(inname, HyoutaUtils::IO::GetFileName(originalpath));
    if (HyoutaUtils::IO::Exists(std::string_view(inname))
        == HyoutaUtils::IO::ExistsResult::DoesExist) {
        std::lock_guard lock(*jobConfig.JobsLock);
        job.TextStatus = std::format("File or directory at {} already exists, cancelling.", inname);
        return ResultType::Failure;
    }
    if (!HyoutaUtils::IO::Move(originalpath, inname, false)) {
        std::lock_guard lock(*jobConfig.JobsLock);
        job.TextStatus = std::format("Moving to {} failed, cancelling.", inname);
        return ResultType::Failure;
    }

    std::string outname = GenerateOutputName(inname, videoInfo->GetVideoId(buffer));

    std::vector<std::string> args;
    args.push_back("-i");
    args.push_back(inname);
    args.push_back("-codec");
    args.push_back("copy");
    args.push_back("-map");
    args.push_back("0");
    args.push_back("-f");
    args.push_back("segment");
    args.push_back("-segment_times");
    args.push_back(splitTimes);
    args.push_back("-segment_start_number");
    args.push_back("1");
    args.push_back("-reset_timestamps");
    args.push_back("1");
    args.push_back(outname);

    std::string exampleOutname = EvaluateOutputName(outname, 0);
    bool existedBefore = HyoutaUtils::IO::Exists(std::string_view(exampleOutname))
                         == HyoutaUtils::IO::ExistsResult::DoesExist;

    {
        auto diskLock = jobConfig.ExpensiveDiskIO.WaitForFreeSlot(cancellationToken);
        if (cancellationToken.IsCancellationRequested()) {
            return ResultType::Cancelled;
        }

        {
            std::lock_guard lock(*jobConfig.JobsLock);
            job.TextStatus = "Splitting...";
        }
        StallWrite(jobConfig,
                   exampleOutname,
                   HyoutaUtils::IO::GetFilesize(std::string_view(inname)).value_or(0),
                   cancellationToken,
                   ShouldStallWriteRegularFile,
                   [&](std::string status) {
                       std::lock_guard lock(*jobConfig.JobsLock);
                       job.TextStatus = std::move(status);
                   });
        int retval = RunProgram(
            "ffmpeg_split.exe", args, [](std::string_view sv) {}, [](std::string_view sv) {});
        if (retval != 0) {
            std::lock_guard lock(*jobConfig.JobsLock);
            job.TextStatus = std::format("ffmpeg_split failed with return value {}", retval);
            return ResultType::Failure;
        }
    }

    {
        std::lock_guard lock(*jobConfig.JobsLock);
        job.TextStatus = "Done!";
        job.JobStatus = VideoJobStatus::Finished;
    }
    return ResultType::Success;
}

ResultType FFMpegSplitJob::Run(JobConfig& jobConfig, TaskCancellation& cancellationToken) {
    return RunSplitJob(*this, jobConfig, cancellationToken);
}

std::string FFMpegSplitJob::GenerateOutputFilename() {
    std::array<char, 256> buffer;
    std::string_view fullPath = VideoInfo->GetVideoId(buffer);
    return GenerateOutputFilenameInternal(fullPath);
}

std::unique_ptr<IVideoJob> FFMpegSplitJob::Clone() const {
    auto clone = std::make_unique<FFMpegSplitJob>();
    clone->TextStatus = this->TextStatus;
    clone->JobStatus = this->JobStatus;
    clone->HasBeenValidated = this->HasBeenValidated;
    clone->VideoInfo = this->VideoInfo ? this->VideoInfo->Clone() : nullptr;
    clone->JobStartTimestamp = this->JobStartTimestamp;
    clone->JobFinishTimestamp = this->JobFinishTimestamp;
    clone->Notes = this->Notes;
    clone->SplitTimes = this->SplitTimes;
    return clone;
}
} // namespace VodArchiver
